halo2_proofs/tests/compress_selectors.rs

```diff
@@ -0,0 +1,421 @@
+
+        let (a0, b0, c0) = cs.raw_multiply(&mut layouter, || x.map(|x| (x, x, x * x)))?;
+        cs.copy(&mut layouter, a0, b0)?;
+
+        let (a1, b1, c1) = cs.raw_multiply(&mut layouter, || y.map(|y| (y, y, y * y)))?;
+        cs.copy(&mut layouter, a1, b1)?;
+
+        let (a2, b2, c2) = cs.raw_add(&mut layouter, || {
+            x.zip(y).map(|(x, y)| (x * x, y * y, x * x + y * y))
+        })?;
+        cs.copy(&mut layouter, a2, c0)?;
+        cs.copy(&mut layouter, b2, c1)?;
+
+        let (a3, b3, c3) = cs.raw_add(&mut layouter, || {
+            x.zip(y)
+                .map(|(x, y)| (x * x + y * y, consty, x * x + y * y + consty))
+        })?;
+        cs.copy(&mut layouter, a3, c2)?;
+        cs.expose_public(&mut layouter, b3, 0)?;
+
+        cs.expose_public(&mut layouter, c3, 1)?;
+
+        Ok(())
+    }
+}
+
+fn test_mycircuit(
+    vk_keygen_compress_selectors: bool,
+    pk_keygen_compress_selectors: bool,
+) -> Result<(), halo2_proofs::plonk::Error> {
+    let engine = PlonkEngineConfig::new()
+        .set_curve::<G1Affine>()
+        .set_msm(H2cEngine::new())
+        .build();
+    let k = 4;
+    let circuit: MyCircuit<Fr> = MyCircuit {
+        x: Value::known(Fr::one()),
+        y: Value::known(Fr::one()),
+        constant: Fr::one(),
+    };
+
+    // Setup
+    let mut rng = BlockRng::new(OneNg {});
+    let params = ParamsKZG::<Bn256>::setup(k, &mut rng);
+    let verifier_params = params.verifier_params();
+    let vk = keygen_vk_custom(&params, &circuit, vk_keygen_compress_selectors)?;
+    let pk = keygen_pk_custom(&params, vk.clone(), &circuit, pk_keygen_compress_selectors)?;
+
+    // Proving
+    let instances = vec![vec![Fr::one(), Fr::from_u128(3)]];
+    let instances_slice: &[&[Fr]] = &(instances
+        .iter()
+        .map(|instance| instance.as_slice())
+        .collect::<Vec<_>>());
+
+    let mut transcript = Blake2bWrite::<_, G1Affine, Challenge255<_>>::init(vec![]);
+    create_proof_with_engine::<KZGCommitmentScheme<Bn256>, ProverSHPLONK<'_, Bn256>, _, _, _, _, _>(
+        engine,
+        &params,
+        &pk,
+        &[circuit],
+        &[instances_slice],
+        &mut rng,
+        &mut transcript,
+    )?;
+    let proof = transcript.finalize();
+
+    // Verify
+    let mut verifier_transcript =
+        Blake2bRead::<_, G1Affine, Challenge255<_>>::init(proof.as_slice());
+    let strategy = SingleStrategy::new(verifier_params);
+
+    verify_proof::<KZGCommitmentScheme<Bn256>, VerifierSHPLONK<'_, Bn256>, _, _, _>(
+        &params,
+        &vk,
+        strategy,
+        &[instances_slice],
+        &mut verifier_transcript,
+    )
+    .map_err(halo2_proofs::plonk::Error::Backend)
+}
+
+#[test]
+fn test_success() {
+    // vk & pk keygen both WITH compress
+    assert!(test_mycircuit(true, true).is_ok());
+
+    // vk & pk keygen both WITHOUT compress
+    assert!(test_mycircuit(false, false).is_ok());
+}
+
+#[should_panic]
```

Comment:
Does this test panic from an out of bounds array access?

Reply:
Yes, exactly.
Fyi, the test panics with following error message.
```
index out of bounds: the len is 2 but the index is 2
```